zebra-state/src/service.rs

```diff
@@ -1554,6 +1550,44 @@ impl Service<ReadRequest> for ReadStateService {
                 .map(|join_result| join_result.expect("panic in ReadRequest::BestChainBlockHash"))
                 .boxed()
             }
+
+            // Used by get_block_template RPC.
+            #[cfg(feature = "getblocktemplate-rpcs")]
+            ReadRequest::ChainInfo => {
+                let timer = CodeTimer::start();
+
+                let state = self.clone();
+                let latest_non_finalized_state = self.latest_non_finalized_state();
+
+                // # Performance
+                //
+                // Allow other async tasks to make progress while concurrently reading blocks from disk.
+                let span = Span::current();
+                tokio::task::spawn_blocking(move || {
+                    span.in_scope(move || {
+                        let expected_difficulty =
+                            read::tip(latest_non_finalized_state.best_chain(), &state.db)
+                                .map(|tip| {
+                                    read::difficulty::relevant_chain_difficulty(
+                                        &latest_non_finalized_state,
+                                        &state.db,
+                                        tip,
+                                        state.network,
+                                    )
+                                })
+                                .unwrap_or(None);
```

Comment:
Actually, I think I was a bit confused by how the function was called.

This should work fine, but it would be clearer if we called `read::best_tip()` in `read::difficulty::relevant_chain_difficulty()`.

We'll also need a comment explaining why it is always correct to do this, even if blocks are added concurrently. The reasons are the same as the block locator hashes:
https://github.com/ZcashFoundation/zebra/blob/6446e0ec1bd42d46c115a1dc41fa78955f4aed30/zebra-state/src/service/read/find.rs#L162-L174

Reply:
What is the big deal if we just leave it as it is now that you understood whats going on ?

Reply:
If we document why it works, then we don't need to make any code changes.

```suggestion
                            // # Correctness
                            //
                            // It is ok to do these lookups using multiple database calls. Finalized state updates
                            // can only add overlapping blocks, and block hashes are unique across all chain forks.
                            // 
                            // If there is a large overlap between the non-finalized and finalized states,
                            // where the finalized tip is above the non-finalized tip,
                            // Zebra is receiving a lot of blocks, or this request has been delayed for a long time.
                            //
                            // In that case, the `getblocktemplate` RPC will return an error because Zebra
                            // is not synced to the tip. That check happens before the RPC makes this request.
                            read::tip(latest_non_finalized_state.best_chain(), &state.db)
                                .map(|tip| {
                                    read::difficulty::relevant_chain_difficulty(
                                        &latest_non_finalized_state,
                                        &state.db,
                                        tip,
                                        state.network,
                                    )
                                })
                                .unwrap_or(None);
```